_cleanup_stale_progress_for_issue deletes active shepherd's progress file

## Bug

`_cleanup_stale_progress_for_issue()` in `ShepherdContext.__post_init__` (`loom-tools/src/loom_tools/shepherd/context.py:54-87`) unconditionally deletes all progress files for the same issue with a different `task_id`, without checking whether the other shepherd is still alive.

## Reproduction

1. Start a shepherd manually: `./.loom/scripts/loom-shepherd.sh 389 -m`
2. While the builder phase is running (15+ minutes), a second shepherd is spawned for the same issue by the orchestration layer
3. The second shepherd's `__post_init__` calls `_cleanup_stale_progress_for_issue()`, which finds `shepherd-<first-task-id>.json` with a matching issue number but different task_id, and deletes it
4. The first shepherd's subsequent `report_milestone()` calls fail with:
   ```
   [ERROR] No progress file found for task <first-task-id>
   [ERROR] Run 'report-milestone.sh started --task-id <first-task-id> --issue N' first
   ```

## Observed behavior

In a real run of issue #389 in the `sphere` repo:

- Shepherd `81be8db` started at 21:30:12, ran all phases successfully, merged PR #394
- Shepherd `f727dd5` was spawned at 21:52:09 (22 minutes later) for the same issue
- `f727dd5`'s init deleted `shepherd-81be8db.json`
- `81be8db` logged `[ERROR] No progress file found` at judge completion, merge completion, and final completion
- `f727dd5` is still sitting as `status: working` with a stale heartbeat from 21:52:11

## Expected behavior

`_cleanup_stale_progress_for_issue` should check `last_heartbeat` freshness before deleting. `agent-wait-bg.sh` sends heartbeats every 60s (`DEFAULT_HEARTBEAT_INTERVAL`), so a heartbeat less than ~120-300s old strongly indicates a live shepherd. The cleanup should skip files with fresh heartbeats instead of unconditionally deleting them.

## Suggested fix

In `context.py:_cleanup_stale_progress_for_issue`, before `progress_file.unlink()`, add a heartbeat freshness check:

```python
# Don't delete files with fresh heartbeats — another shepherd is likely still running
last_hb = data.get("last_heartbeat", "")
if last_hb:
    try:
        age = elapsed_seconds(last_hb)
        if age < HEARTBEAT_STALE_THRESHOLD:  # e.g., 300s
            logger.info(
                "Skipping progress file %s — heartbeat is fresh (%ds old)",
                progress_file.name, age,
            )
            continue
    except (ValueError, OverflowError):
        pass
```

## Severity

Medium — progress tracking breaks (milestone events lost) but actual work (PR creation, review, merge) completes successfully.

--- a/loom-tools/src/loom_tools/shepherd/context.py
+++ b/loom-tools/src/loom_tools/shepherd/context.py
@@ -14,6 +14,7 @@ from loom_tools.common.github import gh_issue_view, gh_list
 from loom_tools.common.paths import LoomPaths, NamingConventions
 from loom_tools.common.repo import find_repo_root
 from loom_tools.common.state import read_json_file
+from loom_tools.common.time_utils import elapsed_seconds
 from loom_tools.shepherd.config import ShepherdConfig
 from loom_tools.shepherd.errors import (
     IssueBlockedError,
@@ -22,6 +23,13 @@ from loom_tools.shepherd.errors import (
     IssueNotFoundError,
 )
 from loom_tools.shepherd.labels import LabelCache, remove_issue_label
+
+# Heartbeats fresher than this (in seconds) indicate a live shepherd.
+# agent-wait-bg.sh sends heartbeats every 60s, so 300s gives ~5 missed
+# heartbeats before we consider the shepherd dead.
+_HEARTBEAT_FRESH_THRESHOLD = int(
+    os.environ.get("LOOM_HEARTBEAT_STALE_THRESHOLD", "300")
+)
 
 
 @dataclass
@@ -61,6 +69,10 @@ class ShepherdContext:
         When a new shepherd starts for an issue that already has a progress
         file (from a previous crashed/orphaned run), remove it to prevent
         stale data from interfering with the new run.
+
+        Files with a fresh heartbeat (< ``_HEARTBEAT_FRESH_THRESHOLD``
+        seconds old) are left intact because they likely belong to a
+        shepherd that is still actively running.
         """
         logger = logging.getLogger(__name__)
         if not self._paths.progress_dir.is_dir():
@@ -78,6 +90,26 @@ class ShepherdContext:
             # Don't remove our own progress file (matched by task_id)
             if data.get("task_id") == self.config.task_id:
                 continue
+
+            # Don't remove files with fresh heartbeats — another shepherd
+            # is likely still running.
+            last_hb = data.get("last_heartbeat", "")
+            if last_hb:
+                try:
+                    age = elapsed_seconds(last_hb)
+                    if age < _HEARTBEAT_FRESH_THRESHOLD:
+                        logger.info(
+                            "Skipping progress file %s for issue #%s — "
+                            "heartbeat is fresh (%ds old, threshold %ds)",
+                            progress_file.name,
+                            issue,
+                            age,
+                            _HEARTBEAT_FRESH_THRESHOLD,
+                        )
+                        continue
+                except (ValueError, OverflowError):
+                    # Unparseable timestamp — treat as stale
+                    pass
 
             logger.info(
                 "Removing stale progress file %s for issue #%s (task_id: %s)",

--- a/loom-tools/tests/shepherd/test_context.py
+++ b/loom-tools/tests/shepherd/test_context.py
@@ -4,13 +4,14 @@ from __future__ import annotations
 
 import json
 import logging
+from datetime import datetime, timedelta, timezone
 from pathlib import Path
 from unittest.mock import MagicMock, patch
 
 import pytest
 
 from loom_tools.shepherd.config import ExecutionMode, ShepherdConfig
-from loom_tools.shepherd.context import ShepherdContext
+from loom_tools.shepherd.context import ShepherdContext, _HEARTBEAT_FRESH_THRESHOLD
 from loom_tools.shepherd.errors import IssueBlockedError, IssueClosedError, IssueIsEpicError
 
 
@@ -117,6 +118,87 @@ class TestCleanupStaleProgressForIssue:
         finally:
             # Restore permissions for cleanup
             progress_dir.chmod(0o755)
+
+
+class TestCleanupHeartbeatFreshness:
+    """Tests for heartbeat freshness check in _cleanup_stale_progress_for_issue()."""
+
+    def test_fresh_heartbeat_preserves_file(self, tmp_path: Path) -> None:
+        """Progress file with a fresh heartbeat is NOT removed."""
+        progress_dir = tmp_path / ".loom" / "progress"
+        progress_dir.mkdir(parents=True)
+        fresh_ts = (datetime.now(timezone.utc) - timedelta(seconds=60)).isoformat()
+        _write_progress(
+            progress_dir,
+            "shepherd-live123.json",
+            {"issue": 42, "task_id": "live123", "last_heartbeat": fresh_ts},
+        )
+
+        _make_context(tmp_path, issue=42, task_id="new5678")
+
+        assert (progress_dir / "shepherd-live123.json").exists()
+
+    def test_stale_heartbeat_removes_file(self, tmp_path: Path) -> None:
+        """Progress file with a stale heartbeat IS removed."""
+        progress_dir = tmp_path / ".loom" / "progress"
+        progress_dir.mkdir(parents=True)
+        stale_ts = (
+            datetime.now(timezone.utc) - timedelta(seconds=_HEARTBEAT_FRESH_THRESHOLD + 60)
+        ).isoformat()
+        _write_progress(
+            progress_dir,
+            "shepherd-dead123.json",
+            {"issue": 42, "task_id": "dead123", "last_heartbeat": stale_ts},
+        )
+
+        _make_context(tmp_path, issue=42, task_id="new5678")
+
+        assert not (progress_dir / "shepherd-dead123.json").exists()
+
+    def test_no_heartbeat_removes_file(self, tmp_path: Path) -> None:
+        """Progress file with no last_heartbeat field is removed (backward compat)."""
+        progress_dir = tmp_path / ".loom" / "progress"
+        progress_dir.mkdir(parents=True)
+        _write_progress(
+            progress_dir,
+            "shepherd-old123.json",
+            {"issue": 42, "task_id": "old123"},
+        )
+
+        _make_context(tmp_path, issue=42, task_id="new5678")
+
+        assert not (progress_dir / "shepherd-old123.json").exists()
+
+    def test_unparseable_heartbeat_removes_file(self, tmp_path: Path) -> None:
+        """Progress file with an invalid heartbeat timestamp is treated as stale."""
+        progress_dir = tmp_path / ".loom" / "progress"
+        progress_dir.mkdir(parents=True)
+        _write_progress(
+            progress_dir,
+            "shepherd-bad123.json",
+            {"issue": 42, "task_id": "bad123", "last_heartbeat": "not-a-timestamp"},
+        )
+
+        _make_context(tmp_path, issue=42, task_id="new5678")
+
+        assert not (progress_dir / "shepherd-bad123.json").exists()
+
+    def test_fresh_heartbeat_logs_skip_message(self, tmp_path: Path, caplog) -> None:
+        """Skipping a file with a fresh heartbeat logs an info message."""
+        progress_dir = tmp_path / ".loom" / "progress"
+        progress_dir.mkdir(parents=True)
+        fresh_ts = (datetime.now(timezone.utc) - timedelta(seconds=30)).isoformat()
+        _write_progress(
+            progress_dir,
+            "shepherd-live456.json",
+            {"issue": 42, "task_id": "live456", "last_heartbeat": fresh_ts},
+        )
+
+        with caplog.at_level(logging.INFO):
+            _make_context(tmp_path, issue=42, task_id="new5678")
+
+        assert "Skipping progress file" in caplog.text
+        assert "heartbeat is fresh" in caplog.text
 
 
 class TestReportMilestone:

